### PR TITLE
Add trait alias support in rustdoc

### LINKED ISSUE #57595: "trait objects are not yet implemented" ICE in rustdoc
Running `cargo doc` on 1aim/asn1@aae670cca45d8632abbd390ab18df97d89becf0e (current master) causes an ICE while running rustdoc on the `serde_asn1` crate:

```
thread 'rustc' panicked at 'not yet implemented: trait objects are not yet implemented', src/librustdoc/visit_ast.rs:550:17
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:70
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:58
             at src/libstd/panicking.rs:200
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:215
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:478
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:385
   6: std::panicking::begin_panic_fmt
             at src/libstd/panicking.rs:340
   7: rustdoc::visit_ast::RustdocVisitor::visit_item
             at src/librustdoc/visit_ast.rs:550
   8: rustdoc::visit_ast::RustdocVisitor::visit_mod_contents
             at src/librustdoc/visit_ast.rs:248
   9: rustdoc::visit_ast::RustdocVisitor::visit_item
             at src/librustdoc/visit_ast.rs:458
  10: rustdoc::visit_ast::RustdocVisitor::visit_mod_contents
             at src/librustdoc/visit_ast.rs:248
  11: rustdoc::visit_ast::RustdocVisitor::visit
             at src/librustdoc/visit_ast.rs:81
  12: rustdoc::core::run_core::{{closure}}::{{closure}}
             at src/librustdoc/core.rs:546
  13: <std::thread::local::LocalKey<T>>::with
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/librustc_driver/driver.rs:1329
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/librustc/ty/context.rs:1955
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/librustc/ty/context.rs:1922
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/librustc/ty/context.rs:1856
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/librustc/ty/context.rs:1921
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/librustc/ty/context.rs:1954
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/librustc/ty/context.rs:1910
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/thread/local.rs:299
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/thread/local.rs:245
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/librustc/ty/context.rs:1902
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/thread/local.rs:299
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libstd/thread/local.rs:245
  14: rustc::ty::context::TyCtxt::create_and_enter
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/librustc/ty/context.rs:1894
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/librustc/ty/context.rs:1933
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/librustc/ty/context.rs:1260
  15: rustc_driver::driver::phase_3_run_analysis_passes
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/librustc_driver/driver.rs:1234
  16: <scoped_tls::ScopedKey<T>>::set
             at src/librustdoc/core.rs:491
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/librustc_driver/driver.rs:65
             at /cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-0.1.2/src/lib.rs:155
  17: rustdoc::core::run_core
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/librustc_driver/driver.rs:64
             at src/librustdoc/core.rs:394
  18: syntax::with_globals
             at src/librustdoc/lib.rs:428
             at /cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-0.1.2/src/lib.rs:155
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libsyntax/lib.rs:111
             at /cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-0.1.2/src/lib.rs:155
             at /rustc/2fadb0a16c8737a45746e95df9138912590ed8ad/src/libsyntax/lib.rs:110
```

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #57595.

r? @QuietMisdreavus

### CODE REVIEW

File: src/librustdoc/clean/mod.rs [hunk 1]
```diff
@@ -1868,13 +1871,41 @@ impl Clean<Item> for doctree::Trait {
...
+pub struct TraitAlias {
+    pub generics: Generics,
+    pub bounds: Vec<GenericBound>,
+    pub is_spotlight: bool,
+}
+
+impl Clean<Item> for doctree::TraitAlias {
+    fn clean(&self, cx: &DocContext) -> Item {
+        let attrs = self.attrs.clean(cx);
+        let is_spotlight = attrs.has_doc_flag("spotlight");
```
Comment (Manishearth) [inline src/librustdoc/clean/mod.rs]: We should check if spotlight actually works for trait impls using an aliased trait.

So, we should see what happens for:

```rust
#[doc(spotlight)]
trait Foo {}
trait Bar = Foo;
struct Ty;
impl Bar for Ty;
```

and 

```rust
trait Foo {}
#[doc(spotlight)]
trait Bar = Foo;
struct Ty;
impl Bar for Ty;
```

File: src/librustdoc/html/render.rs [hunk 2]
```diff
@@ -1908,6 +1911,7 @@ impl fmt::Display for AllTypes {
         print_entries(f, &self.derives, "Derive Macros", "derives")?;
         print_entries(f, &self.functions, "Functions", "functions")?;
         print_entries(f, &self.typedefs, "Typedefs", "typedefs")?;
+        print_entries(f, &self.trait_aliases, "Trait Aliases", "trait-alias")?;
```
Comment (Manishearth) [inline src/librustdoc/html/render.rs]: trait-aliases, to be consistent.

(should there be a hyphen at all?)

**Review by Manishearth — APPROVED** [review]
r=me

File: src/librustdoc/clean/mod.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs]: You can't implement a trait alias directly. So I guess that spotlight doesn't make sense for a trait alias. I'll remove spotlight "support" for them.

File: src/librustdoc/html/render.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render.rs]: It's a css class so I guess it's fine.

Comment (Manishearth) [reply src/librustdoc/html/render.rs]: It's nice to be consistent, also: y'all have theming support so the css classes do matter, if only a little bit

**Comment by GuillaumeGomez** [comment]
Updated.

**Pull request merged**